### ISSUE TITLE
Use an HTTP proxy as a capture device instead of Playwright's network interception?

### ISSUE THREAD

**Issue by matteocargnelutti** [harvard-lil/scoop#4]
> As discussed w/ @leppert 

## Problem(s) identified:
- **Playwright's network interception doesn't give us a full picture of what is happening at network level, which poses capture fidelity challenges.**
  - Some exchanges are never caught.
  - [Response bodies cannot be streamed](https://github.com/microsoft/playwright/issues/17199)
  - We do not have access to the raw HTTP status line and headers. Worse, [duplicate headers seem to get accidentally merged](https://github.com/microsoft/playwright/issues/3184).
- **To compensate for that, we currently rely on:**
  - Out-of-browser capture for exchanges that need to be streamed. This forces us to write our own browser scripts to retrieve relevant resources, since `browsertrix-behaviors` uses in-browser fetch.
  - Compensate for missing information using our _"best guess"_.
  
## Suggested solution:
- **Using an HTTP proxy hooked to Chrome as a capture device.**
  - **Pros:** 
    - This would allow us to access raw HTTP responses, and therefore write higher-fidelity WARCs.
    - Assuming they can be streamed, this would also allow us to use in-browser fetch for capturing additional resources, and therefore integrate `browsertrix-behaviors` _(we may be able to compensate for CORS limitations via Chrome flags, if desirable)_
  - **Cons:**
    - Added complexity _(self-signed certs, moving parts, ...)_
    - Playwright's interpreted HTTP responses are easy to add to WARCs and do not require additional filtering. It is unclear to us at this stage if responses caught via an HTTP proxy would require additional processing / filtering. 

---

### Main:
- [x] Replace Playwright's network interception by an HTTP proxy
  - [ ] HTTP parser need documentation
- [x] Remove out-of-browser capture phase?

### Browser scripts:
- [x] Remove Mischief-specific browser scripts
- [x] Implements `browsertrix-behavior` (behind individual option flags)

### Limits:
> Adapt existing if necessary
- [x] Limit capture time in seconds, and set timeouts for each stage.
- [x] Limit total size of the resources collection in (kilo)bytes.
- [x] Allow to keep partially captured resources.

### Misc:
> Adapt existing if necessary
- [x] Full page screenshots, added to the archive as `file:///screenshot.png`.

### Dealing with the risks associated with permanent one-way transformations:
- [ ] **Idea:** Adding a `HAR` (or equivalent) file with raw HTTP exchanges to the `.wacz` files, so WARCs and CDX can be re-computed / re-interpreted later. 
  - [ ] **Additionally:** Adding logic for re-digesting a `HAR` with Mischief

### Export formats:
> Adapt existing if necessary 
- [x] `.warc`
- [x] `.warc.gz`
- [x] `.wacz`  
  - [ ] _(Error management to solidify)_ 
  - [ ] _(To document)_ 
- [ ] `.wacz` (signed)

### Security:
> Adapt existing if necessary 
 - [x] New browser context for each capture (incognito mode)
 - [x] Only allow `http:` or `https:` as principal capture urls
 - [ ] Since HTTPS cert verification now happens at proxy level, make sure we understand the ins and outs of this switch in security model
- [ ] Prevent harmful redirects _(i.e: redirect to `file://`)_
 - [ ] _(Nice to have/ TBD)_ Since we now have programmatic control over the layer that handles SSL checks, could we capture and record SSL info in the WARC 

### Capture / Playback challenges:
- [ ] Integrate [replayweb.page's capture-side rewrite logic](https://github.com/webrecorder/archiveweb.page/blob/main/src/recorder.js#L952). 
  - Could be it's own shared repo we contribute to (to discuss w/ Ilya)
- [ ] Think-through _"hooks"_ architecture to allow for adding edge case-specific logic to Mischief without adding to its core.
- [ ] Playback challenges: 
  - [ ] What happens when `alert()` or `prompt()` is triggered? 
  - [x] Generic audio / video tags
  - [ ] Twitter videos
  - [x] TikTok videos
  - [ ] YouTube videos
  - [ ] YouTube streams
  - [x] Instagram videos
  - [ ] Facebook videos
  - [ ] Web Sockets?
  - [ ] Service Workers? _(Disable for capture?)_

### Other considerations:
- [x] Adding a capture of the DOM, as a trace of how the browser interpreted this ensemble at the time of capture?
  - [ ] `<replay-web-page>` injects logic in it when we're downloading it.   
- [x] Look into using the Chrome Debugging Protocol instead on an HTTP Proxy for intercepting network traffic.
  - **Conclusions:** Equivalent to Playwright's Network interception in terms of capabilities and how _"close to the metal"_ this actually is. Worth mentioning: unable to stream incomplete responses, missing status line, finicky redirect logic, protocol listed is incorrect ...
 
---

## Post-prototype goals
Evaluate results, see what additional Perma-specific behaviors need to be integrated.
If solution is suitable: establish a roadmap to make this tool production-ready: 
- code robustness 
- error reporting
- unit and integration testing
- documentation
- packaging 
- ...

**Comment by rudreshtrivedi** [comment]
@matteocargnelutti mentioned link is not accessible https://github.com/orgs/harvard-lil/projects/7/views/1

**Comment by matteocargnelutti** [comment]
Hi there! 

This is an issue that @leppert and I used when working on an initial prototype of what would become Scoop. 
The project has changed _a lot_ since then, and I am not sure there is anything helpful here. 
If you would like to read more about how we designed Scoop and the choices we've made along the way, maybe this blog post will help: https://lil.law.harvard.edu/blog/2023/04/13/scoop-witnessing-the-web/ 

Cheers,
